### PR TITLE
[CPU] GQA supports attention scores output

### PR DESCRIPTION
**Pull request by derdeljan-msft**
### Description
1. Add optional output to CPU impl of GQA op for storing attention scores (QK). Buffer is of shape (B, N, S, T) and can either be fp16 or fp32, depending on the type of other inputs
2. Add `qk_output` attribute to GQA, which controls if attention scores should be saved before or after softmax is applied
3. Add unit tests to cover this use case
4. Added asserts on other EPs if this feature is used

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 1]
```diff
@@ -1127,6 +1127,10 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
               "Use a smooth factor in softmax.",
               AttributeProto::INT,
               static_cast<int64_t>(-1))
+        .Attr("output_post_softmax_qk",
```
Comment (kunal-vaishnavi) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc]: There are [many cases](https://github.com/onnx/onnx/pull/6501#discussion_r1975912197) where the QK output is needed. Can we generalize this attribute name to be similar to the ONNX attention op?

**Comment by xadupre** [comment]
I just finished the implementation of Attention(23) on CPU (float only): https://github.com/microsoft/onnxruntime/pull/25156. This operator supports GQA and all the options related to qk matmul output. Then I'll add support float16 and CUDA.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 2]
```diff
@@ -1200,6 +1204,11 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
                 "(k-v buffer), it is of length max_sequence_length... otherwise of length past_sequence_length +"
                 "kv_sequence_length.",
                 "T")
+        .Output(3,
```
Comment (gramalingam) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1247]: The type-and-shape-inference method should be extended to propagate the shape for this optional output. It would also be useful to check in that method that the output is specified if-and-only-if the corresponding attribute value is 0 or 1.

**Review by aciddelgado — DISMISSED** [review]
Suggest adding an error check for when this feature is attempted to be used on other EP's.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 1 above)
Comment (derdeljan-msft) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc]: Sure, but should we still have -1, 0 and 1 as the options (for now)?

Comment (fs-eire) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc]: do we need this attribute at all? My understanding is that ORT should be able to figure out whether to output or not by looking at whether output[3] exists.

Comment (derdeljan-msft) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc]: The idea behind this attribute was to allow the user to specify which version of QK should be written to the output, right now it supports before softmax (0) or after softmax (1). Finer granulation can be implemented, if necessary, I don't need it for my use case though.

Comment (derdeljan-msft) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc]: Renamed the attribute and introduced enum class with valid values for better readability.

File: onnxruntime/contrib_ops/cpu/bert/attention_common.h [hunk 3]
```diff
@@ -53,6 +53,12 @@ enum AttentionKernelType {
   AttentionKernel_Default
 };
 
+enum class QKOutputType : int {
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cpu/bert/attention_common.h:56]: Can you update the MHA op implementation to use this class as well?

File: onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.cc [hunk 4]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "contrib_ops/cpu/bert/group_query_attention_helper.h"
+
+namespace onnxruntime {
+namespace contrib {
+namespace group_query_attention_helper {
+
+Status CheckCPUSpecificFeatures(const std::string& epName, int num_outputs, int qk_output) {
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.cc]: I think we can rename this method to `CheckQKOutput`. We can also remove having the EP name and just say that `output_qk` is not supported.

File: onnxruntime/contrib_ops/cpu/bert/attention_common.h (hunk 3 above)
Comment (derdeljan-msft) [reply onnxruntime/contrib_ops/cpu/bert/attention_common.h:56]: As far I as I see, MHA always outputs QK before the softmax and after the masking, based on whether the output buffer was provided, it doesn't have the corresponding attribute to control this.

File: onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.cc (hunk 4 above)
Comment (derdeljan-msft) [reply onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.cc]: Done

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc [hunk 5]
```diff
@@ -109,6 +109,10 @@ Status GroupQueryAttention<T>::ComputeInternal(OpKernelContext* context) const {
   parameters.do_rotary = do_rotary_;
   parameters.rotary_interleaved = rotary_interleaved_;
 
+  ORT_RETURN_IF_ERROR(group_query_attention_helper::CheckQKOutput(
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc]: The current GQA CUDA implementation will never be able to have a QK output. GQA CUDA uses either flash attention or memory efficient attention. Neither kernel supports returning the QK output.

Side note: This is why GQA is not used in models like Whisper that require the QK output.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 6]
```diff
@@ -1205,10 +1240,15 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
...
+                "output_qk",
+                "Values of QK matrix multiplication, either before or after SoftMax noramlization",
```
Comment (kunal-vaishnavi) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc]:
```suggestion
                "Values of QK matrix multiplication, either before or after softmax normalization",
```

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc (hunk 5 above)
Comment (derdeljan-msft) [reply onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc]: I see, should I remove the assert then or is it fine to stay "just-in-case", as @aciddelgado suggested to add asserts for other EPs?

Comment (kunal-vaishnavi) [reply onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc]: The assert will be useful if an unfused implementation is added in the future that can have a QK output. The assert can stay but adding a comment summarizing the above info may be useful.

Comment (derdeljan-msft) [reply onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc]: Done

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 7]
```diff
@@ -1127,6 +1158,10 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
               "Use a smooth factor in softmax.",
               AttributeProto::INT,
               static_cast<int64_t>(-1))
+        .Attr("qk_output",
+              "Output values of QK matrix multiplication before (0) or after (1) softmax normalization. Default value is -1 (don't output).",
```
Comment (kunal-vaishnavi) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1162]: Can you define the list of valid values here? For example, what values can be used and what they represent?

Comment (kunal-vaishnavi) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc]: Can you define the list of valid values here? What values can be used and what do they represent?

File: onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h [hunk 8]
```diff
@@ -151,8 +156,9 @@ class GQAAttentionBase {
                              const bool past_present_share_buffer,                 // whether present key and value share the same buffer
                              const bool packed_qkv,                                // whether Q, K, V are packed
                              const bool is_prompt,                                 // whether it is prompt
-                             ThreadPool* tp,                                       // thread pool
-                             AllocatorPtr allocator) const {                       // allocator for temporary buffer
+                             T* output_qk,
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h]: Let's insert this parameter after `present_key` to keep all of the `T*` output parameters together and add a comment to say what it is for.

File: onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h [hunk 9]
```diff
@@ -312,13 +323,6 @@ class GQAAttentionBase {
             }
           }
 
-          if (use_smooth_softmax_ || head_sink != nullptr) {
-            float sink = (head_sink != nullptr) ? static_cast<float>(head_sink[head_index]) : 0.0f;
-            ComputeSmoothSoftmaxInplace(output_softmax + start_offset, static_cast<int>(window_size), sink, nullptr);
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h:317]: I don't think the computation for softmax need to be moved below. The computation should be able to stay here and the `QKOutputType::BEFORE_SOFTMAX` can move locations as needed.

Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h:317]: I don't think the computation for softmax need to be moved below. The computation should be able to stay here and the computation for `QKOutputType::BEFORE_SOFTMAX` can move locations as needed.

Comment (derdeljan-msft) [reply onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h:317]: It needs to be moved bellow because we want to first apply the attention mask and then output QK before softmax.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 7 above)
Comment (derdeljan-msft) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc]: But it is already listed here: before(0), after(1) or don't output (-1)

File: onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h (hunk 8 above)
Comment (derdeljan-msft) [reply onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h]: Done

File: onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h [hunk 10]
```diff
@@ -80,8 +84,7 @@ class GQAAttentionBase {
     // Compute the attention score.
     bool gqa_mlas_supported = MlasGQASupported<T>(CblasNoTrans, CblasTrans) &&
                               MlasGQASupported<T>(CblasNoTrans, CblasNoTrans);
-    size_t bytes = SafeInt<size_t>(batch_size) * num_heads_ * sequence_length * seqlen_present_kv_cache *
-                   (gqa_mlas_supported ? sizeof(T) : sizeof(float));
+    size_t bytes = SafeInt<size_t>(batch_size) * num_heads_ * sequence_length * seqlen_present_kv_cache * (gqa_mlas_supported ? sizeof(T) : sizeof(float));
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h:88]: I think output data type is fixed in onnx graph. Not decided by whether MLAS support it or not.

File: onnxruntime/contrib_ops/cpu/bert/attention_common.h [hunk 11]
```diff
@@ -53,6 +53,12 @@ enum AttentionKernelType {
...
+  NO_OUTPUT = -1,
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/attention_common.h]: nit: suggest to use 0, 1, 2 instead of -1, 0, 1 since 0 usually means no, and -1 means invalid.

File: onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h (hunk 10 above)
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h:88]: why seqlen_present_kv_cache instead of total_sequence_length? 

Normally, the qk or softmax output has shape like (batch_size, q_num_heads, q_sequence_length, total_sequence_length) where total_sequence_length = past_sequence_length + kv_sequence_length.

File: onnxruntime/test/python/transformers/test_gqa_cpu.py [hunk 12]
```diff
@@ -1591,12 +1702,26 @@ def parity_check_gqa_prompt_no_buff(
 
     head_sink = get_custom_head_sink(config.num_heads, torch_type=torch_type) if config.has_head_sink else None
 
+    output_qk = (
+        torch.zeros(
+            config.batch_size,
+            config.num_heads,
+            config.kv_sequence_length,
```
Comment (tianleiwu) [inline onnxruntime/test/python/transformers/test_gqa_cpu.py:1709]: This shape does not looks right. It shall be like (batch_size, num_head, q_sequence_length, total_sequence_length).

File: onnxruntime/test/python/transformers/test_gqa_cpu.py [hunk 13]
```diff
@@ -1873,10 +2018,24 @@ def parity_check_gqa_past(
         else None
     )
 
+    output_qk = (
```
Comment (tianleiwu) [inline onnxruntime/test/python/transformers/test_gqa_cpu.py:2021]: Do not need this. You can see other outputs like out has no zeros tensor passed to gqa_past_func

Comment (derdeljan-msft) [reply onnxruntime/test/python/transformers/test_gqa_cpu.py:2021]: It is required since we expect the user to provide the actual buffer with allocated memory, in which the op will write the output.

File: onnxruntime/contrib_ops/cpu/bert/attention_common.h (hunk 11 above)
Comment (derdeljan-msft) [reply onnxruntime/contrib_ops/cpu/bert/attention_common.h]: Done

File: onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h (hunk 10 above)
Comment (derdeljan-msft) [reply onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h:88]: This logic is unchanged, here we actually allocate the temp buffer for performing this computation. Later, if the user provided the actual buffer in which to store output QK, this temp buffer will be copied to, and potentially casted to the appropriate data type.

File: onnxruntime/test/python/transformers/test_gqa_cpu.py (hunk 12 above)
Comment (derdeljan-msft) [reply onnxruntime/test/python/transformers/test_gqa_cpu.py:1709]: For prompt case, q_sequence_length is the same as kv_sequence_length from what I've seen

File: onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h (hunk 10 above)
Comment (derdeljan-msft) [reply onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h:88]: Update output shape to be total_sequence_length. The allocation logic here remains unchanged (I am not sure why seqlen_present_cache instead of total_sequence_length is being allocated here, I am not really familiar with MLAS kernel for this one)

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 14]
```diff
@@ -336,19 +339,47 @@ void BaseGroupQueryAttentionTypeAndShapeInference(ONNX_NAMESPACE::InferenceConte
           updateOutputShape(ctx, 2, present_shape);
         }
       }
+
+      if (output_qk_index >= 0) {
+        const bool did_supply_qk_buffer = ctx.hasOutput(output_qk_index);
+        const int64_t qk_output_type = getAttribute(ctx, "qk_output", 0);
```
Comment (gramalingam) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc]: nit: safer to use `QKOutputType::NO_OUTPUT` instead of 0 in this line and lines below. (For some reason, I thought it was -1 for no-output.)

Comment (derdeljan-msft) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc]: It used to be, I changed it to be 0, 1, 2 instead of -1, 0, 1. Makes sense to use the enum here for the checks as well, updated.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 15]
```diff
@@ -1205,10 +1244,15 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
                 "(k-v buffer), it is of length max_sequence_length... otherwise of length past_sequence_length +"
                 "kv_sequence_length.",
                 "T")
+        .Output(3,
+                "output_qk",
+                "Values of QK matrix multiplication, either before or after softmax noramlization",
```
Comment (kunal-vaishnavi) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc]:
```suggestion
                "Values of QK matrix multiplication, either before or after softmax normalization",
```

File: onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h [hunk 16]
```diff
@@ -398,6 +398,37 @@ Status CheckCustomAttentionInputs(const T* position_ids,
...
+
+  if (qk_output != static_cast<int>(QKOutputType::NO_OUTPUT) && output_qk == nullptr) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "qk_output attribute was configured but output buffer was not provided");
+  }
+
+  return Status::OK();
+}
+
+inline Status CheckQKOutput(int num_outputs, int qk_output) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h]: nit: Maybe rename this function to `CheckNoQkOutput`

File: onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h [hunk 17]
```diff
@@ -398,6 +398,37 @@ Status CheckCustomAttentionInputs(const T* position_ids,
...
+  if (num_outputs > 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "output_qk optional output is not supported on");
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h]: nit: remove on. Same below

**Comment by tianleiwu** [comment]
A related work is to update symbolic shape inference: https://github.com/microsoft/onnxruntime/blob/9de58ac7a3d18d6ae7f7ae502b3f91361067f1b5/onnxruntime/python/tools/symbolic_shape_infer.py#L2548
I think it could be in another PR if needed in the future.

It would be nice to have same qk output modes like the onnx Attention op. We may update it later to make them consistent.

Overall, it looks good to me.

**Review by tianleiwu — APPROVED** [review]

**Comment by derdeljan-msft** [comment]
> A related work is to update symbolic shape inference:
> 
> https://github.com/microsoft/onnxruntime/blob/9de58ac7a3d18d6ae7f7ae502b3f91361067f1b5/onnxruntime/python/tools/symbolic_shape_infer.py#L2548
> 
> 
> I think it could be in another PR if needed in the future.
> It would be nice to have same qk output modes like the onnx Attention op. We may update it later to make them consistent.
> 
> Overall, it looks good to me.

Makes sense, opened an issue to track this item https://github.com/microsoft/onnxruntime/issues/25412

**Review by kunal-vaishnavi — APPROVED** [review]

**Review by tianleiwu — APPROVED** [review]
